File: desktop/app/src/renderer/app-state/message-manager.js
class MessageList {
  constructor(id, updateCallback) {
    this.id = id;
    this.list = [];
    this.unread = 0;
    this.updateCallback = updateCallback;
  }

  push(message) {
    this.unread++;
    this.list.push(message);
  }

  readAll() {
    if (this.unread) {
      this.unread = 0;
      this.updateCallback();
    }
  }

  pushReceive(message) {
    this.push({
      id: +new Date(),
      time: new Date(),
      content: message,
      type: 'recv',
    });
  }
  pushSend(message) {
    this.push({
      id: +new Date(),
      time: new Date(),
      content: message,
      type: 'send',
    });
  }
  pushSystem(message) {
    this.push({
      id: +new Date(),
      time: new Date(),
      content: message,
      type: 'sys',
    });
  }
}

export default class MessageManager {
  constructor(updateCallback) {
    this.messageList = {};
    this.updateCallback = updateCallback;
  }

  addNewMessage(id, message) {
    this.getList(id).push(message);
  }

  getList(id) {
    if (!this.messageList[id]) {
      this.messageList[id] = new MessageList(id, this.updateCallback);
    }
    return this.messageList[id];
  }

  getUnread(id) {
    return this.getList(id).unread;
  }

  readAll(id) {
    this.getList(id).readAll();
  }

  getMessages(id) {
    return this.getList(id).list;
  }
}

File: android/app/src/main/java/cn/jackq/messenger/message/MessageManager.java
package cn.jackq.messenger.message;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class MessageManager {
    private Map<String, MessageList> store = new HashMap<>();

    public void addNewMessage(String user, Message message) {
    }

    public List<Message> getMessages(String id) {
        return getMessageList(id).getList();
    }

    public int getUnread(String id) {
        return getMessageList(id).getUnread();
    }

    public void readAll(String id) {
        getMessageList(id).readAll();
    }

    public void addMessage(String id, Message message) {
        getMessageList(id).addMessage(message);
    }

    private MessageList getMessageList(String id) {
        if (!this.store.containsKey(id)) {
            this.store.put(id, new MessageList());
        }
        return this.store.get(id);
    }

    private MessageManager() {
    }

    private static MessageManager list = new MessageManager();

    public static MessageManager get() {
        return list;
    }

}
